src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs

```diff
@@ -306,7 +306,7 @@ impl<'tcx> RegionInferenceContext<'tcx> {
                     region_naming,
                 );
 
-                outlives_suggestion.collect_constraint(fr, outlived_fr);
+                outlives_suggestion.collect_constraint(fr, outlived_fr, errors_buffer.last_mut());
```

Comment:
Do we know if this will always work on incremental builds as expected? I fear that we might give the notes to the incorrect error by doing it this way (and action at a distance can be scary!)

Reply:
I will check again, but I believe this is always called after an outlives error is emitted (or rather buffered), so we should be good.

I agree that this is action at a distance (which I don't like), but the distance is not that great, as the immediately preceding function call is the one that emits the error. The problem is that there isn't really a natural place in that function to emit the intermediate help message, and it would be annoying to pass the `OutlivesSuggestionBuilder` through the arguments of everything it calls.

Reply:
This is resolved in the most recent commits. Basically, I refactored the error reporting code to return the `DiagnosticBuilder` for the generated error, which we can then add a suggestion to before buffering.